src/mac/region/RegionUS915.c

```diff
@@ -141,43 +141,49 @@ static LoRaMacStatus_t ComputeNext125kHzJoinChannel( uint8_t* newChannelIndex )
     uint16_t channelMaskRemaining;
     uint8_t findAvailableChannelsIndex[8] = { 0 };
     uint8_t availableChannels = 0;
+    uint8_t startIndex = NvmCtx.JoinChannelGroupsCurrentIndex;
 
     // Null pointer check
     if( newChannelIndex == NULL )
     {
         return LORAMAC_STATUS_PARAMETER_INVALID;
     }
 
-    // Current ChannelMaskRemaining, two groups per channel mask. For example Group 0 and 1 (8 bit) are ChannelMaskRemaining 0 (16 bit), etc.
-    currentChannelsMaskRemainingIndex = (uint8_t) NvmCtx.JoinChannelGroupsCurrentIndex / 2;
+    do {
+        // Current ChannelMaskRemaining, two groups per channel mask. For example Group 0 and 1 (8 bit) are ChannelMaskRemaining 0 (16 bit), etc.
+        currentChannelsMaskRemainingIndex = (uint8_t) startIndex / 2;
 
-    // For even numbers we need the 8 LSBs and for uneven the 8 MSBs
-    if( ( NvmCtx.JoinChannelGroupsCurrentIndex % 2 ) == 0 )
-    {
-        channelMaskRemaining = ( NvmCtx.ChannelsMaskRemaining[currentChannelsMaskRemainingIndex] & 0x00FF );
-    }
-    else
-    {
-        channelMaskRemaining = ( ( NvmCtx.ChannelsMaskRemaining[currentChannelsMaskRemainingIndex] >> 8 ) & 0x00FF );
-    }
+        // For even numbers we need the 8 LSBs and for uneven the 8 MSBs
+        if( ( startIndex % 2 ) == 0 )
+        {
+            channelMaskRemaining = ( NvmCtx.ChannelsMaskRemaining[currentChannelsMaskRemainingIndex] & 0x00FF );
+        }
+        else
+        {
+            channelMaskRemaining = ( ( NvmCtx.ChannelsMaskRemaining[currentChannelsMaskRemainingIndex] >> 8 ) & 0x00FF );
+        }
 
 
-    if( FindAvailable125kHzChannels( findAvailableChannelsIndex, channelMaskRemaining, &availableChannels ) == LORAMAC_STATUS_PARAMETER_INVALID )
-    {
-        return LORAMAC_STATUS_PARAMETER_INVALID;
-    }
+        if( FindAvailable125kHzChannels( findAvailableChannelsIndex, channelMaskRemaining, &availableChannels ) == LORAMAC_STATUS_PARAMETER_INVALID )
+        {
+            return LORAMAC_STATUS_PARAMETER_INVALID;
+        }
 
-    // Choose randomly a free channel 125kHz
-    *newChannelIndex = findAvailableChannelsIndex[randr( 0, ( availableChannels - 1 ) )];
+        if (availableChannels) {
+            // Choose randomly a free channel 125kHz
+            *newChannelIndex = (startIndex * 8) + findAvailableChannelsIndex[randr( 0, ( availableChannels - 1 ) )];
+        }
+        startIndex++;
+        if (startIndex > 7) startIndex = 0;
+    } while(availableChannels == 0 && startIndex != NvmCtx.JoinChannelGroupsCurrentIndex);
 

```

Comment:
Before we process the following code, a verification of `if( availableChannels > 0 )` would be useful. Should be then:
```
    if( availableChannels > 0 )
    {
        NvmCtx.JoinChannelGroupsCurrentIndex = startIndex++;

        if( NvmCtx.JoinChannelGroupsCurrentIndex > 7 )
        {
            // Start again from group 0
            NvmCtx.JoinChannelGroupsCurrentIndex = 0;
        }
        return LORAMAC_STATUS_OK;
    }

    return LORAMAC_STATUS_PARAMETER_INVALID;
```